File: src/data/listings.js
import { marketFor } from '../lib/market.js'

// Listings are generated for whatever place the user is browsing, seeded by
// that place's coordinates, so the same city always yields the same homes.
// ponytail: generated inventory. The shape matches what an MLS/IDX feed would
// return, so swapping in a real feed is a change to one function.

export const photos = [
  'https://images.unsplash.com/photo-1600596542815-ffad4c1539a9?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1600585154340-be6161a56a0c?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1600607687939-ce8a6c25118c?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1600566753086-00f18fb6b3ea?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1600573472591-ee6b68d14c68?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1583608205776-bfd35f0d9f83?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1570129477492-45c003edd2be?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1564013799919-ab600027ffc6?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1576941089067-2de3c901e126?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1598228723793-52759bba239c?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1512917774080-9991f1c4c750?w=600&h=400&fit=crop',
  'https://images.unsplash.com/photo-1600047509807-ba8f99d2cdde?w=600&h=400&fit=crop'
]

export const propertyTypes = ['house', 'condo', 'townhouse']

// Where the house number goes. Most of continental Europe and much of Asia put
// it after the street name; the anglosphere puts it first.
const numberAfterStreet = new Set([
  'DE', 'AT', 'CH', 'NL', 'BE', 'LU', 'SE', 'NO', 'DK', 'FI', 'IS', 'PL', 'CZ', 'SK',
  'HU', 'RO', 'BG', 'RS', 'HR', 'SI', 'GR', 'IT', 'ES', 'PT', 'TR', 'RU', 'UA', 'EE',
  'LV', 'LT', 'BR', 'AR', 'CL', 'CO', 'PE', 'MX', 'ID', 'VN', 'NP', 'JP', 'CN', 'TW', 'KR'
])

function formatAddress(number, street, countryCode) {
  return numberAfterStreet.has(countryCode) ? `${street} ${number}` : `${number} ${street}`
}

function seededRandom(seed) {
  let s = Math.abs(Math.trunc(seed)) % 2147483646 || 1
  return function () {
    s = (s * 16807) % 2147483647
    return (s - 1) / 2147483646
  }
}

// Two places 100m apart should not produce identical inventory.
function seedFor(place) {
  return Math.round((place.lat + 90) * 1e4) * 100003 + Math.round((place.lng + 180) * 1e4)
}

const pick = (rand, arr) => arr[Math.floor(rand() * arr.length)]

// How wide to scatter homes when Overpass gave us no streets. Kept near the
// centroid, which is always on land, because a wider scatter drops homes into
// the sea for any coastal city.
// ponytail: no land mask. A real one needs coastline polygons or a water query;
// the street anchors are on land by construction, so this only has to cover the
// degraded path, which the UI already labels as approximate.
function spreadFor(place) {
  const [s, w, n, e] = place.bbox || []
  const span = place.bbox ? Math.min(n - s, e - w) : 0.06
  return Math.min(Math.max(span * 0.5, 0.02), 0.04)
}

/**
 * @param place  from geo.js — needs lat/lng/name/countryCode
 * @param streets optional real streets from Overpass; falls back to the place name
 * @param mode   'sale' | 'rent'
 */
export function generateListings(place, streets = [], mode = 'sale', count = 60) {
  const market = marketFor(place.countryCode)
  const rand = seededRandom(seedFor(place) + (mode === 'rent' ? 7919 : 0))
  const listings = []
  // Base value of a mid-market home in this country, before size/type.
  const base = 400000 * market.level

  for (let i = 0; i < count; i++) {
    const type = pick(rand, propertyTypes)
    const anchor = streets.length ? pick(rand, streets) : place
    const street = streets.length ? anchor.name : place.name
    // Without real streets every home would anchor to the one city point and
    // the map would show a single blob, so fall back to the place's own bbox.
    const jitter = streets.length ? 0.012 : spreadFor(place)

    let beds, baths, sqft, sizeFactor
    if (type === 'house') {
      beds = Math.floor(rand() * 4) + 2
      baths = Math.floor(rand() * 3) + 1
      sqft = Math.floor(rand() * 2000) + 1200
      sizeFactor = 1.6
    } else if (type === 'condo') {
      beds = Math.floor(rand() * 3) + 1
      baths = Math.max(1, Math.round(beds / 2))
      sqft = Math.floor(rand() * 800) + 450
      sizeFactor = 0.8
    } else {
      beds = Math.floor(rand() * 3) + 2
      baths = Math.floor(rand() * 2) + 1
      sqft = Math.floor(rand() * 1200) + 800
      sizeFactor = 1.15
    }

    const spread = 0.6 + rand() * 0.9
    const value = base * sizeFactor * spread
    // Rent is the same asset priced by yield, so a market's rents track its
    // sale prices instead of being invented separately.
    const price = mode === 'rent'
      ? roundTo(value * market.grossYield / 12, 10)
      : roundTo(value, 1000)

    listings.push({
      id: `${place.id}-${mode}-${i}`,
      address: formatAddress(Math.floor(rand() * 220) + 1, street, market.countryCode),
      neighborhood: streets.length ? street : place.name,
      city: place.name,
      countryCode: market.countryCode,
      currency: market.currency,
      imperial: market.imperial,
      mode,
      price,
      beds,
      baths,
      sqft,
      type,
      lat: anchor.lat + (rand() - 0.5) * jitter,
      lng: anchor.lng + (rand() - 0.5) * jitter,
      year: Math.floor(rand() * 60) + 1965,
      photo: photos[i % photos.length],
      photos: [photos[i % photos.length], photos[(i + 3) % photos.length], photos[(i + 7) % photos.length]],
      listedDaysAgo: Math.floor(rand() * 30) + 1,
      refNumber: `R${100000 + Math.floor(rand() * 899999)}`
    })
  }

  return listings
}

function roundTo(n, step) {
  return Math.max(step, Math.round(n / step) * step)
}

File: kmp/shared/src/commonMain/kotlin/com/nulljosh/roost/Listings.kt
package com.nulljosh.roost

import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.roundToLong

// Ported from src/data/listings.js. Listings are generated for whatever
// place the user is browsing, seeded by that place's coordinates, so the
// same city always yields the same homes.
// ponytail: generated inventory, no real streets (streetsNear/Overpass not
// ported -- see Geo.kt), so every listing anchors to the place centroid
// with the fallback jitter spread, same as the web app's degraded path.

val PROPERTY_TYPES = listOf("house", "condo", "townhouse")

private val NUMBER_AFTER_STREET = setOf(
    "DE", "AT", "CH", "NL", "BE", "LU", "SE", "NO", "DK", "FI", "IS", "PL", "CZ", "SK",
    "HU", "RO", "BG", "RS", "HR", "SI", "GR", "IT", "ES", "PT", "TR", "RU", "UA", "EE",
    "LV", "LT", "BR", "AR", "CL", "CO", "PE", "MX", "ID", "VN", "NP", "JP", "CN", "TW", "KR",
)

private fun formatAddress(number: Int, street: String, countryCode: String): String =
    if (countryCode in NUMBER_AFTER_STREET) "$street $number" else "$number $street"

/** Park-Miller LCG, same constants as seededRandom() in listings.js -- must
 *  produce the identical sequence for the same seed as the web app. */
private class SeededRandom(seed: Double) {
    private var s: Long = (abs(seed.toLong()) % 2147483646L).let { if (it == 0L) 1L else it }
    fun next(): Double {
        s = (s * 16807) % 2147483647
        return (s - 1).toDouble() / 2147483646.0
    }
}

private fun seedFor(place: Place): Double =
    kotlin.math.round((place.lat + 90) * 1e4) * 100003 + kotlin.math.round((place.lng + 180) * 1e4)

private fun <T> pick(rand: SeededRandom, list: List<T>): T = list[(rand.next() * list.size).toInt()]

private fun spreadFor(): Double = 0.06 * 0.5 // no bbox ported, so always the no-bbox default span

private fun roundTo(n: Double, step: Double): Double = max(step, (n / step).roundToLong() * step)

data class Listing(
    val id: String,
    val address: String,
    val neighborhood: String,
    val city: String,
    val countryCode: String,
    val currency: String,
    val imperial: Boolean,
    val mode: String,
    val price: Double,
    val beds: Int,
    val baths: Int,
    val sqft: Int,
    val type: String,
    val lat: Double,
    val lng: Double,
    val year: Int,
    val listedDaysAgo: Int,
    val refNumber: String,
)

fun generateListings(place: Place, mode: String = "sale", count: Int = 60): List<Listing> {
    val market = marketFor(place.countryCode)
    val rand = SeededRandom(seedFor(place) + if (mode == "rent") 7919.0 else 0.0)
    val listings = mutableListOf<Listing>()
    val base = 400000 * market.level

    repeat(count) { i ->
        val type = pick(rand, PROPERTY_TYPES)
        val street = place.name
        val jitter = spreadFor()

        val beds: Int; val baths: Int; val sqft: Int; val sizeFactor: Double
        when (type) {
            "house" -> {
                beds = floor(rand.next() * 4).toInt() + 2
                baths = floor(rand.next() * 3).toInt() + 1
                sqft = floor(rand.next() * 2000).toInt() + 1200
                sizeFactor = 1.6
            }
            "condo" -> {
                beds = floor(rand.next() * 3).toInt() + 1
                baths = max(1, (beds / 2.0).roundToLong().toInt())
                sqft = floor(rand.next() * 800).toInt() + 450
                sizeFactor = 0.8
            }
            else -> {
                beds = floor(rand.next() * 3).toInt() + 2
                baths = floor(rand.next() * 2).toInt() + 1
                sqft = floor(rand.next() * 1200).toInt() + 800
                sizeFactor = 1.15
            }
        }

        val spread = 0.6 + rand.next() * 0.9
        val value = base * sizeFactor * spread
        val price = if (mode == "rent") roundTo(value * market.grossYield / 12, 10.0) else roundTo(value, 1000.0)

        listings.add(
            Listing(
                id = "${place.id}-$mode-$i",
                address = formatAddress(floor(rand.next() * 220).toInt() + 1, street, market.countryCode),
                neighborhood = place.name,
                city = place.name,
                countryCode = market.countryCode,
                currency = market.currency,
                imperial = market.imperial,
                mode = mode,
                price = price,
                beds = beds,
                baths = baths,
                sqft = sqft,
                type = type,
                lat = place.lat + (rand.next() - 0.5) * jitter,
                lng = place.lng + (rand.next() - 0.5) * jitter,
                year = floor(rand.next() * 60).toInt() + 1965,
                listedDaysAgo = floor(rand.next() * 30).toInt() + 1,
                refNumber = "R${100000 + floor(rand.next() * 899999).toInt()}",
            ),
        )
    }

    return listings
}
